osu.Game.Rulesets.Taiko/Skinning/LegacyTaikoScroller.cs

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Screens.Play;
+using osu.Game.Skinning;
+using osuTK;
+
+namespace osu.Game.Rulesets.Taiko.Skinning
+{
+    public class LegacyTaikoScroller : CompositeDrawable
+    {
+        public LegacyTaikoScroller()
+        {
+            RelativeSizeAxes = Axes.Both;
+        }
+
+        [BackgroundDependencyLoader(true)]
+        private void load(GameplayBeatmap gameplayBeatmap)
+        {
+            if (gameplayBeatmap != null)
+                ((IBindable<JudgementResult>)LastResult).BindTo(gameplayBeatmap.LastJudgementResult);
+        }
+
+        private bool passing;
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            LastResult.BindValueChanged(result =>
+            {
+                var r = result.NewValue;
+
+                passing = r == null || (r.Judgement.AffectsCombo && r.Type > HitResult.Miss);
+
+                foreach (var sprite in InternalChildren.OfType<ScrollerSprite>())
+                    sprite.Passing = passing;
+            }, true);
+        }
+
+        public Bindable<JudgementResult> LastResult = new Bindable<JudgementResult>();
+
+        protected override void Update()
+        {
+            base.Update();
+
+            while (true)
+            {
+                float? additiveX = null;
+
+                foreach (var sprite in InternalChildren)
+                {
+                    // add the x coordinates and perform re-layout on all sprites as spacing may change with gameplay scale.
+                    sprite.X = additiveX ??= sprite.X - (float)Time.Elapsed * 0.1f;
+
+                    additiveX += sprite.DrawWidth - 1;
+
+                    if (sprite.X + sprite.DrawWidth < 0)
+                        sprite.Expire();
+                }
+
+                var last = InternalChildren.LastOrDefault();
+
+                // only break from this loop once we have saturated horizontal space completely.
+                if (last != null && last.ScreenSpaceDrawQuad.TopRight.X >= ScreenSpaceDrawQuad.TopRight.X)
+                    break;
+
+                AddInternal(new ScrollerSprite
+                {
+                    Passing = passing
```

Comment:
This is set too early - when I did the repro scenario described in the earlier comment in the test scene, this crashed with a nullref:

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at osu.Framework.Graphics.TransformableExtensions.PopulateTransform[TValue,TEasing,TThis](TThis t, Transform`3 transform, TValue newValue, Double duration, TEasing& easing)
   at osu.Framework.Graphics.TransformableExtensions.MakeTransform[TThis,TEasing,TValue](TThis t, String propertyOrFieldName, TValue newValue, Double duration, TEasing& easing)
   at osu.Framework.Graphics.TransformableExtensions.TransformTo[TThis,TValue,TEasing](TThis t, String propertyOrFieldName, TValue newValue, Double duration, TEasing& easing)
   at osu.Framework.Graphics.TransformableExtensions.FadeTo[T,TEasing](T drawable, Single newAlpha, Double duration, TEasing& easing)
   at osu.Framework.Graphics.TransformableExtensions.FadeIn[T,TEasing](T drawable, Double duration, TEasing& easing)
   at osu.Framework.Graphics.TransformableExtensions.FadeIn[T](T drawable, Double duration, Easing easing)
   at osu.Game.Rulesets.Taiko.Skinning.LegacyTaikoScroller.ScrollerSprite.set_Passing(Boolean value) in D:\Open Source\osu\osu.Game.Rulesets.Taiko\Skinning\LegacyTaikoScroller.cs:line 107
```

Should probably be done on `LoadComplete()`.

Even then the fade duration in the setter is a problem. It should probably fade instantly the first time and then normally after load complete, but even that can look slightly wrong if you miss close to the seam between two sprites (the left one will transition over 200ms, while the right one will spawn already transitioned).

I'm not sure how viable it is but I'd actually probably consider waiting for the texture wrapping framework-side changes and using that instead - it should be a relatively fool-proof method and would get rid of the infinite loop relayouting, too, if it doesn't end up being too expensive performance-wise.